packages/react/menu/src/Menu.tsx

```diff
@@ -621,6 +636,206 @@ const MenuItemIndicator = React.forwardRef((props, forwardedRef) => {
 
 MenuItemIndicator.displayName = ITEM_INDICATOR_NAME;
 
+/* -------------------------------------------------------------------------------------------------
+ * MenuSubMenu
+ * -----------------------------------------------------------------------------------------------*/
+
+const SUB_MENU_NAME = 'MenuSubMenu';
+
+type MenuSubMenuContextValue = {
+  triggerRef: React.RefObject<HTMLDivElement>;
+  contentId: string;
+  focusFirstItem: boolean;
+  onMouseOpen(): void;
+  onKeyboardOpen(): void;
+};
+
+const [SubMenuProvider, useSubMenuContext] = createContext<MenuSubMenuContextValue>(SUB_MENU_NAME);
+
+type MenuSubMenuOwnProps = {
+  open?: boolean;
+  onOpenChange?(open: boolean): void;
+  defaultOpen?: boolean;
+};
+
+const MenuSubMenu: React.FC<MenuSubMenuOwnProps> = (props) => {
+  const { children, open: openProp, defaultOpen, onOpenChange } = props;
+  const triggerRef = React.useRef<HTMLDivElement>(null);
+  const [focusFirstItem, setFocusFirstItem] = React.useState(false);
+  const [renderChildren, setRenderChildren] = React.useState(false);
+  const [open = false, setOpen] = useControllableState({
+    prop: openProp,
+    defaultProp: defaultOpen,
+    onChange: onOpenChange,
+  });
+
+  /**
+   * We defer rendering of child elements until ready, this ensures the following:
+   *
+   * - That measurements are applied correctly to children relative to parent
+   * - That dismissable layers are appended in the correct order when using controlled props
+   */
+  useLayoutEffect(() => {
+    setRenderChildren(true);
+  }, []);
+
+  return (
+    <SubMenuProvider
+      contentId={useId()}
+      triggerRef={triggerRef}
+      focusFirstItem={focusFirstItem}
+      onMouseOpen={React.useCallback(() => {
+        setOpen(true);
+        setFocusFirstItem(false);
+      }, [setOpen])}
+      onKeyboardOpen={React.useCallback(() => {
+        setOpen(true);
+        setFocusFirstItem(true);
+      }, [setOpen])}
+    >
+      <Menu open={open} onOpenChange={setOpen}>
+        {renderChildren && children}
+      </Menu>
+    </SubMenuProvider>
+  );
+};
+
+MenuSubMenu.displayName = SUB_MENU_NAME;
+
+/* -------------------------------------------------------------------------------------------------
+ * MenuSubMenuContent
+ * -----------------------------------------------------------------------------------------------*/
+
+const SUB_MENU_CONTENT_NAME = 'MenuSubMenuContent';
+
+type MenuSubMenuContentOwnProps = Omit<
+  Polymorphic.OwnProps<typeof MenuContent>,
+  'portalled' | 'disableOutsidePointerEvents' | 'disableOutsideScroll'
```

Comment:
Narrowing here because:
- `popper` doesn't currently support relative position afaik, so submenus must always be `portalled`
- The way pointer events interact here is fundamental to submenus working as expected, I can't think of a case for users wanting to change this.